### PR TITLE
10316 Deferred `bracket`

### PR DESCRIPTION
**Pull request by exarkun**
## Scope and purpose

See https://twistedmatrix.com/trac/ticket/10316

Additionally, this would be useful inside Twisted in refactoring some of disttrial's implementation.  Specifically, my wip for that currently has:

```
        return bracket(
            # Start the worker pool.
            partial(pool.start, reactor, count),
            # Start submitting tests to workers in the pool.  Perhaps repeat
            # the whole test suite more than once, if appropriate for our
            # configuration.
            whileM_(
                partial(runTests, pool, testCases),
                shouldContinue,
            ),
            # Shut down the worker pool.
            pool.join,
        )
```

to implement safe worker process startup, shutdown, and `--until-failure` behavior.

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10316
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [x] I have updated the automated tests and checked that all checks for the PR are green.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
* [x] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.

```
Merge pull request #123 from twisted/4356-branch-name-with-trac-id

Author: exarkun
Reviewer: <github_username>, <github_username_if_more_reviewers>
Fixes: ticket:10316

Long description providing a summary of these changes.
(as long as you wish)
```

### CODE REVIEW

**Comment by adiroiban** [comment]
Changes looks good.
As long as this is private API, it should be ok.
But the tests are failing and I think we need green tests to merge this. Thanks!

**Review by adiroiban — APPROVED** [review]
Let's merge and see how it goes.

This is private API so we can always refactor , bugfix or remove later.

Thanks!

File: src/twisted/trial/_dist/control.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    This is a functional version of C{async with ...} that is convenient to
+    use when setup and teardown are already defined as functions, when a
+    partially applied form of the async context manager is useful, and for
+    other functional-style composition purposes.
+
+    @param first: A no-argument function that may return a Deferred.  It is
+        called first.
+
+    @param between: A no-argument function that may return a Deferred.  It is
```
Comment (twm) [inline src/twisted/trial/_dist/control.py:32]: Move this paragraph down below `@param last` to match argument order?

File: src/twisted/trial/_dist/test/test_bracket.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        assert_that(
+            result,
+            instance_of(SomeException),
+        )
+        assert_that(
+            actions,
+            equal_to(["first", "between", "last"]),
+        )
+
+    def test_success_with_failing_last(self) -> None:
```
Comment (twm) [inline src/twisted/trial/_dist/test/test_bracket.py:81]: Naming convention: 

```suggestion
    def test_successWithFailingLast(self) -> None:
```

And other tests below.

File: src/twisted/trial/_dist/test/test_bracket.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+@define
+class BracketTests:
+    """
+    Tests for ``bracket``.
+    """
+
+    case: SynchronousTestCase
+    wrap_success: Callable
+    wrap_failure: Callable
```
Comment (twm) [inline src/twisted/trial/_dist/test/test_bracket.py:28]: Naming convention: 

```suggestion
    wrapFailure: Callable
```

File: src/twisted/trial/_dist/test/test_bracket.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+    def test_success(self) -> None:
+        """
+        ``bracket`` calls ``first`` then ``between`` then ``last`` and returns a
+        ``Deferred`` that fires with the result of ``between``.
```
Comment (twm) [inline src/twisted/trial/_dist/test/test_bracket.py:33]: This doesn't appear correct: `bracket()` returns an awaitable, not necessarily a `Deferred`.

**Review by twm — APPROVED** [review]
@exarkun This looks good to land with some minor changes. It's been sitting here a while, though &mdash; are you still pushing it forward?

**Comment by exarkun** [comment]
Thanks for the feedback, all. I don't think I'll pursue this further.

**Pull request closed without merging**